metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -0,0 +1,387 @@
+          case _ => None
+        }
+      }
+    }
+
+    member match {
+      case c: Defn.Class => namesFromTemplate(c.templ)
+      case t: Defn.Trait => namesFromTemplate(t.templ)
+      case o: Defn.Object => namesFromTemplate(o.templ)
+      case e: Defn.Enum => namesFromTemplate(e.templ)
+    }
+  }
+
+  private def renameFileAsMemberAction(
+      uri: String,
+      member: Member
+  ): l.CodeAction = {
+    val className = member.name.value
+    val newUri = newPathFromClass(uri, member).toURI.toString
+    val fileName = uri.toAbsolutePath.filename
+
+    val edits: List[Either[l.TextDocumentEdit, l.ResourceOperation]] = List(
+      Right(new l.RenameFile(uri, newUri))
+    )
+
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(
+      ExtractRenameMember.renameFileAsClassTitle(fileName, className)
+    )
+    codeAction.setKind(l.CodeActionKind.Refactor)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(edits.map(_.asJava).asJava)
+    )
+
+    codeAction
+  }
+
+  private def extractClassAction(
+      uri: String,
+      member: Member,
+      title: String
+  ): l.CodeAction = {
+
+    val range = member.name.pos.toLSP
+
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(title)
+    codeAction.setKind(kind)
+    codeAction.setCommand(
+      ServerCommands.ExtractMemberDefinition.toLSP(
+        List(
+          uri,
+          range.getStart.getLine(): java.lang.Integer,
+          range.getStart.getCharacter(): java.lang.Integer
+        )
+      )
+    )
+
+    codeAction
+  }
+
+  def executeCommand(
+      data: ExtractMemberDefinitionData
+  ): CodeActionCommandResult = {
+    val uri = data.uri
+    val params = data.params
+
+    def isCompanion(member: Member)(candidateCompanion: Member): Boolean = {
+      val differentMemberWithSameName = !candidateCompanion.equals(member) &&
+        candidateCompanion.name.value.equals(member.name.value)
+      member match {
+        case _: Defn.Object => differentMemberWithSameName
+        case _ =>
+          candidateCompanion
+            .isInstanceOf[Defn.Object] && differentMemberWithSameName
+      }
+    }
+
+    val pos = params.getPosition
+    val range = new l.Range(pos, pos)
+    val path = uri.toAbsolutePath
+
+    val opt = for {
+      tree <- trees.get(path)
+      definitions = membersDefinitions(tree)
+      memberDefn <- definitions.find(_.name.pos.toLSP.overlapsWith(range))
+      companion = definitions.find(isCompanion(memberDefn))
+      fileContent = newFileContent(
+        tree,
+        range,
+        memberDefn,
+        companion
+      )
+      newFilePath = newPathFromClass(uri, memberDefn)
+      if !newFilePath.exists
+      newFileMemberRange = new l.Range()
+    } yield (
+      newFilePath,
+      fileContent,
+      memberDefn,
+      companion,
+      newFileMemberRange
+    )
+
+    // from checks already done when proposing the CodeAction
+    // we know that the data in the Option will exist
+    val (
```

Comment:
Can you not do this all in the yield block?

```scala
val opt = for {
      tree <- trees.get(path)
      definitions = membersDefinitions(tree)
      memberDefn <- definitions.find(_.name.pos.toLSP.overlapsWith(range))
      companion = definitions.find(isCompanion(memberDefn))
      fileContent = newFileContent(
        tree,
        range,
        memberDefn,
        companion
      )
      newFilePath = newPathFromClass(uri, memberDefn)
      if !newFilePath.exists
      newFileMemberRange = new l.Range()
    } yield {
      val newFileUri = newFilePath.toURI.toString

      extractClassCommand(
        newFileUri,
        fileContent,
        memberDefn,
        companion
      )
    }
```

you can later do `opt.toIterable.flatten`

Reply:
I need to extract also the `newFileUri`, would something like this be acceptable?

```scala
yield {
      val newFileUri = newFilePath.toURI.toString
      (extractClassCommand(
        newFileUri,
        fileContent,
        memberDefn,
        companion
      ), newFileUri)
    }

    val edits = opt.map(_._1).toList.flatten
    val newFileUri = opt.map(_._2).get
    ```

Reply:
I would put everything into the yield then and return CodeActionCommandResult from it. We can the treat None in this case Future.failed